### PR TITLE
[DAG] select m, sub/add(X, C), X --> sub/add (X, and(C, m))

### LINKED ISSUE #66101: [DAG] Failure to fold select(x, sub(x, c), m) -> sub(x, and(c,m))
https://godbolt.org/z/a1PczEM8a

If we're selecting a subtracting a non-constant we fold the select into a and:
```c
#include <x86intrin.h>
auto masked_select(__m128i a, __m128i b, __m128i x, __m128i y) {
    return _mm_blendv_epi8(a, _mm_sub_epi32(a, b), _mm_cmpgt_epi32(x,y));
}
```
```asm
masked_select(long long __vector(2), long long __vector(2), long long __vector(2), long long __vector(2)): # @masked_select(long long __vector(2), long long __vector(2), long long __vector(2), long long __vector(2))
  pcmpgtd %xmm3, %xmm2
  pand %xmm1, %xmm2
  psubd %xmm2, %xmm0
  retq
```
But for constants this fails, which on x86 can result in a BLENDV instruction, which is never faster than a AND
```c
#include <x86intrin.h>
auto masked_select_const(__m128i a, __m128i x, __m128i y) {
    __m128i b = _mm_set1_epi32(24);
    return _mm_blendv_epi8(a, _mm_sub_epi32(a, b), _mm_cmpgt_epi32(x,y));
}
```
```asm
masked_select_const(long long __vector(2), long long __vector(2), long long __vector(2)): # @masked_select_const(long long __vector(2), long long __vector(2), long long __vector(2))
  movdqa %xmm0, %xmm3
  movdqa .LCPI3_0(%rip), %xmm4 # xmm4 = [4294967272,4294967272,4294967272,4294967272]
  paddd %xmm0, %xmm4
  pcmpgtd %xmm2, %xmm1
  movdqa %xmm1, %xmm0
  blendvps %xmm0, %xmm4, %xmm3
  movaps %xmm3, %xmm0
  retq
```

### PR DESCRIPTION
**Pull request by elhewaty**
- [DAG] Add tests for Folding select m, sub(X, C), X --> sub (X, and(C, m))(NFC)
- [DAG][X86] Fold select m, sub/add(X, C), X --> sub/add (X, and(C, m))
- Fixes: https://github.com/llvm/llvm-project/issues/66101

### CODE REVIEW

**Comment by elhewaty** [comment]
@RKSimon

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 1]
```diff
@@ -11660,6 +11660,24 @@ SDValue DAGCombiner::visitSELECT(SDNode *N) {
     }
   }
 
+  // select m, sub(X, C), X --> sub (X, and(C, m))
```
Comment (topperc) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:11663]: I'm confused. Shouldn't select have a scalar condition. If the condition is a vector it should be ISD::VSELECT

File: llvm/test/CodeGen/X86/vselect.ll [hunk 2]
```diff
@@ -22,7 +22,7 @@ define <2 x i64> @masked_select_const(<2 x i64> %a, <2 x i64> %x, <2 x i64> %y)
 ; SSE41-LABEL: masked_select_const:
 ; SSE41:       # %bb.0:
 ; SSE41-NEXT:    movdqa %xmm0, %xmm3
-; SSE41-NEXT:    movdqa {{.*#+}} xmm4 = [4294967272,4294967272,4294967272,4294967272]
+; SSE41-NEXT:    pmovsxbd {{.*#+}} xmm4 = [4294967272,4294967272,4294967272,4294967272]
```
Comment (topperc) [inline llvm/test/CodeGen/X86/vselect.ll:25]: This test already produces pmovsxbd before your patch.

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp (hunk 1 above)
Comment (RKSimon) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:11663]: These folds need moving to visitVSELECT

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 3]
```diff
@@ -11660,6 +11660,24 @@ SDValue DAGCombiner::visitSELECT(SDNode *N) {
...
+  }
+
+  // select (sext m), (add X, C), X --> (add X, (and C, (sext m))))
+  if (N1.getOpcode() == ISD::ADD && N1.getOperand(0) == N2 && N1->hasOneUse() &&
+      DAG.isConstantIntBuildVectorOrConstantInt(N1.getOperand(1)) && 
+      N0.getScalarValueSizeInBits() == N1.getScalarValueSizeInBits()) {
+    return DAG.getNode(ISD::ADD, DL, N1.getValueType(), N2,
+                       DAG.getNode(ISD::AND, DL, N0.getValueType(), N1.getOperand(1),
+                                   N0));
+  }
```
Comment (RKSimon) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:11679]: You should be able to merge the ADD/SUB folds? However, unless we have test coverage for the SUB case you'll need to drop it and just keep the ADD case.

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 4]
```diff
@@ -11660,6 +11660,24 @@ SDValue DAGCombiner::visitSELECT(SDNode *N) {
     }
   }
 
+  // select m, sub(X, C), X --> sub (X, and(C, m))
+  if (N1.getOpcode() == ISD::SUB && N1.getOperand(0) == N2 && N1->hasOneUse() &&
+      DAG.isConstantIntBuildVectorOrConstantInt(N1.getOperand(1)) &&
+      N0.getScalarValueSizeInBits() == N1.getScalarValueSizeInBits()) {
```
Comment (RKSimon) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:11666]: We need a DAG.ComputeNumSignBits(N0) == N0.getScalarValueSizeInBits() check as well

File: llvm/test/CodeGen/X86/vselect.ll (hunk 2 above)
Comment (RKSimon) [reply llvm/test/CodeGen/X86/vselect.ll:25]: The base commit with the tests needs rebasing (and regenerating) on trunk first

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp (hunk 3 above)
Comment (goldsteinn) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:11679]: The sub case could be `select m, (sub C, X), (sub 0, X)`

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp (hunk 4 above)
Comment (goldsteinn) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:11666]: Shouldn't `sext` guarantee that?

**Comment by RKSimon** [comment]
@elhewaty reverse-ping?

**Comment by elhewaty** [comment]
@RKSimon, I am sorry for my late reply, my hard drive is corrupted, and I lost the repo. I cloned it again. and I will submit the new patch tomorrow.

**Comment by RKSimon** [comment]
Close this and just use #83640 ?

**Pull request closed without merging**